### ISSUE TITLE
340.108 Causes 'Oh no, something has gone wrong'

### ISSUE THREAD

**Issue by threehappypenguins** [MeowIce/nvidia-legacy#8]
I followed the directions to install driver 340.108 for kernel 5.15 and afterwards I get the screen 'Oh no, something has gone wrong'. I was unable to get it to install with TTY, because I couldn't shut stop gdm3 service (it would go to a black screen with the blinking line cursor thing in the top left corner; then I had to hard boot). So I installed the driver outside the OS with command line. It gives me cc warning (which I ignore), then I get this warning:

`WARNING: nvidia-installer was forced to guess the X library path 'usr/lib' and X module path '/usr/lib/xorg/modules'; these path were not queryable from the system. If X fails to find the NVIDIA X driver module, please install the 'pkg-config' utility and the X.Org SDK/development package for your distribution and reinstall the driver.`

It still seems to successfully install. Upon booting up, I get the error and cannot get into the OS.

**Comment by threehappypenguins** [comment]
I learned a little more about what I was doing. Reinstalled the OS and tried again. For one thing, pressing ctrl+alt+F4 (F4 in my case) brings me to TTY, and I didn't realize that I all I had to do after shutting down gdm3 was press ctrl+alt+F4 again to get the command line interface again.

I installed the driver, got the "Oh no!" screen again, and then pressed ctrl+alt+F4 which brought up TTY. I tried reinstalling the driver, and then starting up gdm3 again and still got the "Oh no!" screen. I tried uninstalling gdm3 and installing lightdm, then later sddm, and both got the "Oh no!" screen as well. The only difference is that I actually got a login screen with them and it did the "Oh no!" thing after I tried to log in.

The other thing that is happening is when I execute `nvidia-settings`, I get `Error: The control display is undefined; please run `nvidia-settings --help` for usage information`.

Even after removal of the nvidia driver, and enabling nouveau again, I never recover from getting the "Oh no!" screen and I have to reinstall Ubuntu.

**Comment by MeowIce** [comment]
> I learned a little more about what I was doing. Reinstalled the OS and tried again. For one thing, pressing ctrl+alt+F4 (F4 in my case) brings me to TTY, and I didn't realize that I all I had to do after shutting down gdm3 was press ctrl+alt+F4 again to get the command line interface again.
> 
> I installed the driver, got the "Oh no!" screen again, and then pressed ctrl+alt+F4 which brought up TTY. I tried reinstalling the driver, and then starting up gdm3 again and still got the "Oh no!" screen. I tried uninstalling gdm3 and installing lightdm, then later sddm, and both got the "Oh no!" screen as well. The only difference is that I actually got a login screen with them and it did the "Oh no!" thing after I tried to log in.
> 
> The other thing that is happening is when I execute `nvidia-settings`, I get `Error: The control display is undefined; please run `nvidia-settings --help` for usage information`.
> 
> Even after removal of the nvidia driver, and enabling nouveau again, I never recover from getting the "Oh no!" screen and I have to reinstall Ubuntu.

which Ubuntu version are you running ? Above 22.04 on GNOME version could cause this issue since they have removed the support for legacy drivers and I can not do anything.

**Comment by MeowIce** [comment]
> 22.04.1. The only 22.04 version I can find on the Ubuntu site. Are there earlier versions of 22.04? If there are, I can't find them and don't know what they're called. Just 22.04? Maybe 22.04.0?

https://old-releases.ubuntu.com/releases/

Ubuntu names their releases by yy.mm
22.04 means relased on April 2022.

**Comment by MeowIce** [comment]
> I finally got a little bit somewhere. Firstly, I now understand what a [window manager and a DE (Desktop Environment](https://askubuntu.com/questions/18078/what-is-the-difference-between-a-desktop-environment-and-a-window-manager) is.
> 
> From a fresh install, I tried again, but instead of GNOME, I decided to go with [KDE Plasma](https://computingforgeeks.com/install-kde-plasma-desktop-on-ubuntu-linux/) as my DE (in installed the minimal version).
> 
> After install and reboot, as I suspected, I still got the "Oh no!" screen because the driver seems to be incompatible with current gdm3 as well. So I switched to sddm (for my window manager) with `sudo dpkg-reconfigure gdm3` and then chose sddm since it was already installed. I then restarted the window manager and got the log in screen. I logged in, and the KDE plasma DE loaded fine!
> 
> I opened terminal and opened `nvidia-settings` and the Nvidia setting Gui popped up. But I got the error:
> 
> `You do not appear to be using the NVIDIA X driver. Please edit your X configuration file (just run 'nvidia-xconfig' as root) and restart the X server.`
> 
> When I run `nvidia-xconfig` again, I get:
> 
> ```
> Using X configuration file: "/etc/X11/xorg.conf"
> Backed up file '/etc/X11/xorg.conf' as '/etc/X11/xorg.conf.backup'
> New X configuration file written to '/etc/X11/xorg.conf'
> ```
> 
> It made no difference either. I also tried reinstalling the Nvidia driver (I had to do this in Ubuntu recovery with network and root enabled because on TTY it complained about the Nvidia driver running). This also made no difference. Then I tried `sudo startx` and that's where I'm stuck. At first, I get the "Oh no!" error with the white screen. Then if I open TTY and restart sddm, I get a black screen with a mouse now. I have tried changing to xcfe for my DE and lightdm for my window manager. I try to log in and it loops back to the login screen. I tried KDE Plasma with lightdm and that produces a login loop too. And sddm with xfce produces a black screen with the mouse.
> 
> Here is an example of an error log for sudo `journalctl -xeu lightdm.service`:
> 
> ```
> pam_kwallet5(lightdm-greeter:setcred): pam_kwallet5: pam_sm_setcred
> pam unix(lightdm-greeter:session): session oepened for user lightdm(uid=130) by (uid=0)
> pam_succeed_if(lghtdm:auth): requirement "user ingroup nopasswdlogin" not met by user "threehappypenguins"
> gkr-pam: unable to locate daemon control file
> gkr-pam: stashed password to try later in open session
> pam_kwallet5(lightdm:auth): pam_kwallet5: pam_sm_authenticate
> pam_kwallet5(lightdm:setcred): pam_kwallet5: pam_sm_setcred
> pam_unix(lightdm:session): session opened for user threehappypenguins(uid=1000) by (uid=0)
> Error writing X authority: Failed to open X authority /home/threehappypenguins/.Xauthority: Permission denied
> gkr-pam: gnome-keyring-daemon started properly and unlocked keyring
> pam_kwallet5(lightdm:session): pam_kwallet5: pam_sm_open_session
> Error reading existing Xauthority: Failed to open file "home/sarah/.Xauthority": Permission denied
> pam_succeed_if(lightdm:auth): requirement "user ingroup nopasswdlogin" not met by user "threehappypenguins"
> ```
> `tail -100 ~/.xsession-errors` shows (the last few lines):
> 
> ```
> /usr/bin/startxfce4: X server already running on display :0
> Authorization required, but no authorization protocol specified
> xrdb: Permission denied
> xrdb: Can't open display ':0'
> Authorization required, but no authorization protocol specified
> xfce4-session: Cannot open display:
> Type 'xfce4-session --help' for usage.
> ```
> Then I did `sudo chown $USER:$USER ~/.Xauthority` (replacing $USER with my username). I restarted lightdm (as that was I was running at that moment) and was able to finally log in!
> 
> However, I am back to square one again. When I run `nvidia-settings` in terminal again, I get the "You do not appear to be using the NVIDIA X driver" error again. Then I try `sudo startx` again, and I get the "Oh no!" screen again. So I restart the lightdm service and try to log in, and I'm back to the login loop. `~/.xsession-errors` shows X connection to :0.0 is broken (explicit kill or server shutdown). journalctl for lightdm shows a similar log, but without the permission denied stuff this time.
> 
> Lastly, I checked /var/log/Xorg.0.log just not and I'm getting stuff like:
> 
> ```
> config/udev: Adding input device HDA Nvidia HDMI/DP,pcm=3 (/dev/input/event11)
> No input driver specified, ignoring this device.
> ```
> 
> And there's pcm=7 and 8 too.
> 
> Any idea of what is happening and what to do?

Try Ubuntu Unity 22.04, I think it should work...

**Comment by MeowIce** [comment]
> > I finally got a little bit somewhere. Firstly, I now understand what a [window manager and a DE (Desktop Environment](https://askubuntu.com/questions/18078/what-is-the-difference-between-a-desktop-environment-and-a-window-manager) is.
> > 
> > From a fresh install, I tried again, but instead of GNOME, I decided to go with [KDE Plasma](https://computingforgeeks.com/install-kde-plasma-desktop-on-ubuntu-linux/) as my DE (in installed the minimal version).
> > 
> > After install and reboot, as I suspected, I still got the "Oh no!" screen because the driver seems to be incompatible with current gdm3 as well. So I switched to sddm (for my window manager) with `sudo dpkg-reconfigure gdm3` and then chose sddm since it was already installed. I then restarted the window manager and got the log in screen. I logged in, and the KDE plasma DE loaded fine!
> > 
> > I opened terminal and opened `nvidia-settings` and the Nvidia setting Gui popped up. But I got the error:
> > 
> > `You do not appear to be using the NVIDIA X driver. Please edit your X configuration file (just run 'nvidia-xconfig' as root) and restart the X server.`
> > 
> > When I run `nvidia-xconfig` again, I get:
> > 
> > ```
> > Using X configuration file: "/etc/X11/xorg.conf"
> > Backed up file '/etc/X11/xorg.conf' as '/etc/X11/xorg.conf.backup'
> > New X configuration file written to '/etc/X11/xorg.conf'
> > ```
> > 
> > It made no difference either. I also tried reinstalling the Nvidia driver (I had to do this in Ubuntu recovery with network and root enabled because on TTY it complained about the Nvidia driver running). This also made no difference. Then I tried `sudo startx` and that's where I'm stuck. At first, I get the "Oh no!" error with the white screen. Then if I open TTY and restart sddm, I get a black screen with a mouse now. I have tried changing to xcfe for my DE and lightdm for my window manager. I try to log in and it loops back to the login screen. I tried KDE Plasma with lightdm and that produces a login loop too. And sddm with xfce produces a black screen with the mouse.
> > 
> > Here is an example of an error log for sudo `journalctl -xeu lightdm.service`:
> > 
> > ```
> > pam_kwallet5(lightdm-greeter:setcred): pam_kwallet5: pam_sm_setcred
> > pam unix(lightdm-greeter:session): session oepened for user lightdm(uid=130) by (uid=0)
> > pam_succeed_if(lghtdm:auth): requirement "user ingroup nopasswdlogin" not met by user "threehappypenguins"
> > gkr-pam: unable to locate daemon control file
> > gkr-pam: stashed password to try later in open session
> > pam_kwallet5(lightdm:auth): pam_kwallet5: pam_sm_authenticate
> > pam_kwallet5(lightdm:setcred): pam_kwallet5: pam_sm_setcred
> > pam_unix(lightdm:session): session opened for user threehappypenguins(uid=1000) by (uid=0)
> > Error writing X authority: Failed to open X authority /home/threehappypenguins/.Xauthority: Permission denied
> > gkr-pam: gnome-keyring-daemon started properly and unlocked keyring
> > pam_kwallet5(lightdm:session): pam_kwallet5: pam_sm_open_session
> > Error reading existing Xauthority: Failed to open file "home/sarah/.Xauthority": Permission denied
> > pam_succeed_if(lightdm:auth): requirement "user ingroup nopasswdlogin" not met by user "threehappypenguins"
> > ```
> > `tail -100 ~/.xsession-errors` shows (the last few lines):
> > 
> > ```
> > /usr/bin/startxfce4: X server already running on display :0
> > Authorization required, but no authorization protocol specified
> > xrdb: Permission denied
> > xrdb: Can't open display ':0'
> > Authorization required, but no authorization protocol specified
> > xfce4-session: Cannot open display:
> > Type 'xfce4-session --help' for usage.
> > ```
> > Then I did `sudo chown $USER:$USER ~/.Xauthority` (replacing $USER with my username). I restarted lightdm (as that was I was running at that moment) and was able to finally log in!
> > 
> > However, I am back to square one again. When I run `nvidia-settings` in terminal again, I get the "You do not appear to be using the NVIDIA X driver" error again. Then I try `sudo startx` again, and I get the "Oh no!" screen again. So I restart the lightdm service and try to log in, and I'm back to the login loop. `~/.xsession-errors` shows X connection to :0.0 is broken (explicit kill or server shutdown). journalctl for lightdm shows a similar log, but without the permission denied stuff this time.
> > 
> > Lastly, I checked /var/log/Xorg.0.log just not and I'm getting stuff like:
> > 
> > ```
> > config/udev: Adding input device HDA Nvidia HDMI/DP,pcm=3 (/dev/input/event11)
> > No input driver specified, ignoring this device.
> > ```
> > 
> > And there's pcm=7 and 8 too.
> > 
> > Any idea of what is happening and what to do?
> 
> Try Ubuntu Unity 22.04, I think it should work...

And as I said, Ubuntu had removed support for legacy drivers. Maybe you should try to install 20.04 instead.

**Comment by iseesam** [comment]
I just tried Ubuntu 22.04 (the release from April). Exact same issues and error logs.

Is there any way to know ahead of time which releases support legacy NVIDIA drivers? I don't want to go down to Ubuntu 20 because it's not LTS and I'd like to stick with LTS. I had the same issue with the latest Pop OS as well, so I'm not going to bother with that one. Thinking of Mint, maybe.

**Comment by threehappypenguins** [comment]
Sorry, I understood you as saying that the gnome that came with Ubuntu 22.04.1 removed support for legacy drivers. I didn't know you meant Ubuntu 22.04.1 itself.

I just tried Ubuntu 22.04 (the release from April). Exact same issues and error logs.

Is there any way to know ahead of time which releases support legacy NVIDIA drivers? I don't want to go down to Ubuntu 20 because it's not LTS and I'd like to stick with LTS. I had the same issue with the latest Pop OS as well, so I'm not going to bother with that one. Thinking of Mint, maybe.

**Comment by threehappypenguins** [comment]
Update. I just tried on Ubuntu 20.04.5 and when the desktop manager tries to start, the screen goes black (like the screen is off; not even the "bright" lit up black like when the screen is on), never to recover (not even TTY). In recovery, I looked at `~/.xession-errors` and it showed"

`gnome-session: Fatal IO error 11 (Resource temporarily unavailable) on X server :0.`

Same with xfce (I had tried that afterwards; it will say xfce-session: Fatal IO). There's other errors, lie "Failed to set the background" and "Gtk-WARNING **" Negative content width -3 (allocation 1, extens 2x2) while allocating gadget (node button, owner PowerManagerButton).

I checked Xorg.0.log and everything looks fine; the NVIDIA modules all loaded.

**Comment by MeowIce** [comment]
> Update. I just tried on Ubuntu 20.04.5 and when the window manager tries to start, the screen goes black (like the screen is off; not even the "bright" lit up black like when the screen is on), never to recover (not even TTY). In recovery, I looked at `~/.xession-errors` and it showed"
> 
> `gnome-session: Fatal IO error 11 (Resource temporarily unavailable) on X server :0.`
> 
> Same with xfce (I had tried that afterwards; it will say xfce-session: Fatal IO). There's other errors, like "Failed to set the background" and "Gtk-WARNING **" Negative content width -3 (allocation 1, extens 2x2) while allocating gadget (node button, owner PowerManagerButton).
> 
> I checked Xorg.0.log and everything looks fine; the NVIDIA modules all loaded.

Just upgraded to Ubuntu 22.04 today. Kernel 5.15 is the default one. You might want to try the 5.18 patch. It should work !

**Comment by threehappypenguins** [comment]
> I've just upgraded to Ubuntu 22.04 today. Kernel 5.15 is the default one. You might want to try the 5.18 patch for 5.15 (sounds strange...but it worked for me). It should be alright !

I installed a fresh version of Ubuntu 22.04 (not 22.04.1), downloaded the 5.18 patch (I have the 5.15 kernel) and nvidia.ko fails to build. I am getting:

````
In file included from /tmp/selfgz48978/NVIDIA-Linux-x86_64-340.108/kernel/nv.c:13:
/tmp/selfgz48978/NVIDIA-Linux-x86_64-340.108/kernel/nv-linux.h: At top level:
/tmp/selfgz48978/NVIDIA-Linux-x86_64-340.108/kernel/nv-linux.h:122:10: fatal error: asm/kmap_types.h: No such file or directory
  122 | #include <asm/kmap_types.h>            /* page table entry lookup          */
compilation terminated.
make[2]: *** [scripts/Makefile.build:285: /tmp/selfgz48978/NVIDIA-Linux-x86_64-340.108/kernel/nv.o] Error 1
make[1]: *** [Makefile:1875: /tmp/selfgz48978/NVIDIA-Linux-x86_64-340.108/kernel] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.15.0-25-generic'
NVIDIA: left KBUILD.
nvidia.ko: failed to build!
make: *** [Makefile:197: nvidia.ko] Error 1
```
`

**Comment by MeowIce** [comment]
Try with Kernel 6.0. I have uploaded the patch for 6.0. If problem still exist, try Fedora. I've tested 340.108 6.0 patch on Fedora 37 and everything worked fine.

**Comment by MeowIce** [comment]
> > Try with Kernel 6.0. I have uploaded the patch for 6.0. If problem still exist, try Fedora. I've tested 340.108 6.0 patch on Fedora 37 and everything worked fine.
> 
> Sorry, do you mean the patch 6.0 with my current kernel of 5.15?

Upgrade to Kernel 6.0 then try again, or try Fedora. I did not have kernel panic on Fedora.

**Comment by MeowIce** [comment]
You can also try kelebek333's nvidia legacy on [launchpad](https://launchpad.net/~kelebek333/+archive/ubuntu/nvidia-legacy). It is a just work patch, settings will be lost after logout.

**Comment by threehappypenguins** [comment]
> Upgrade to Kernel 6.0 then try again with patch 6.0, or try Fedora. I did not have kernel panic on Fedora.

I updated my kernel to 6.0 using [these directions](https://www.techrepublic.com/article/install-linux-kernel-ubuntu/). Then tried to install driver and got this error:
```
cc: error: unrecognized command-line option '-ftrivial-auto-var-init=zero'
```
It seems that the default gcc 11.x.x was the issue, so I installed gcc 12 and switched it:

```
sudo apt install gcc-12
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-12 12
```
Then the installation finished. I rebooted, and got the 'Oh no' screen. So I did apt update and upgrade, rebooted, reinstalled the nvidia driver, rebooted again, and still got the 'Oh no' screen. Then I replaced gdm3 with lightdm. The login screen loaded, but once I logged in, got the 'Oh no' screen again. I checked ~/.xsession-errors and it is showing `xrdb: Can't open display ''` and `xhost: unable to open display ""`

I guess next I will have to reinstall Ubuntu 22.04 again in order to have 5.15 kernel again, and then try kelebek333's nvidia legacy (provided it will work with 5.15 and not need 5.18). If that doesn't work, then I'll try upgrading to 5.18 and try again, and if that doesn't work, I guess on to Fedora to try that.

I don't understand why it's not working for me, but it's working for you.

**Comment by threehappypenguins** [comment]
> You can also try kelebek333's nvidia legacy on [launchpad](https://launchpad.net/~kelebek333/+archive/ubuntu/nvidia-legacy). It is a just work patch, settings will be lost after logout.

I just tried kelebek333's patch. I did the following:
```
sudo add-apt-repository ppa:kelebek333/nvidia-legacy
sudo apt update
sudo apt install xorg-modulepath-fix
```
xorg-modulepath-fix seems to also install the driver, and when I rebooted, I never get to the login screen. The screen turns black as I described in a [previous comment](https://github.com/MeowIce/nvidia-legacy/issues/8#issuecomment-1324481292). I also couldn't boot into recovery, so I had to launch shell to purge everything nvidia driver. Before I did that, I checked Xorg.0.log and it shows the nvidia modules and everything loaded. I tried to exit from shell and it went into kernel panic. So I hard booted and then booted up Ubuntu and it booted fine as long as I had purged everything nvidia.

So I don't know what's happening or what I'm doing wrong (or what to do). I guess I have to try Fedora.
